sycl/doc/design/VirtualFunctions.md

```diff
@@ -0,0 +1,428 @@
+all device virtual functions, but their definitions are outlined and linked
+back dynamically based on device capabilities.
+
+For AOT flow, we don't do outlining and dynamic linking, but instead do direct
+cleanup of virtual functions which are incompatible with a target device.
+
+## Design
+
+### Changes to the SYCL header files
+
+New compile-time properties `indirectly_callable` and `calls_indirectly` should
+be implemented in accordance with the corresponding [design document][2]:
+
+- `indirectly_callable` property should lead to emission of
+  `"indirectly-callable"="set"` function attribute, where "set" is a string
+  representation of the property template parameter.
+- `calls_indirectly` property should lead to emission of
+  `"calls-indirectly"="set1,set2"`, where "set1" and "set2" are string
+  representations of the property template parameters.
+
+In order to convert a type to a string, [\__builtin_sycl_unique_stable_name][3]
+could be used.
+
+**TODO**: `calls_indirectly` requires compile-time concatenation of strings.
+Document how it should be done.
+
+`indirectly_callable` property is applied to functions using "custom" (comparing
+to other properties) `SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY` macro. This
+is done to allow implementations to attach some extra attributes alongside the
+property. In particular, functions marked with the macro should be considered
+SYCL device functions and compiler should emit diagnostics if those functions
+do not conform with the SYCL 2020 specification. To achieve that and avoid
+extending FE to parse strings within properties, the aforementioned macro should
+also set `sycl_device` attribute:
+
+```
+#define SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY(SetId)                    \
+  __attribute__((sycl_device)) [[__sycl_detail__::add_ir_attribute_function(   \
+      "indirectly-callable", __builtin_sycl_unique_stable_name(SetId))]]
+```
+
+### Changes to the compiler front-end
+
+Most of the handling for virtual functions happens in middle-end and thanks to
+compile-time properties, no extra work is required to propagate necessary
+information down to passes from headers.
+
+However, we do need to filter out those virtual functions which are not
+considered to be device  as defined by the [extension specification][1], such
+as:
+
+- virtual member functions annotated with `indirectly_callable` compile-time
+  property should be emitted into device code;
+- virtual member function *not* annotated with `indirectly_callable`
+  compile-time property should *not* be emitted into device code;
+
+There is no need to actually check which exact property is applied to a
+function, it is enough to check if `add_ir_attribute_function` attribute was
+applied and that we are in SYCL device mode to decide whether or not a virtual
+function should be emitted into vtable and device code.
+
+**TODO:** any extra diagnostics we would like to emit? Like kernel without
```

Comment:
I'm not sure all of them can be diagnosed by FE, but here's a list of potentially problematic things that I can think of:

- Kernel without `calls_indirectly` property performing virtual function call. (Already mentioned in the doc)
- Kernel with `calls_indirectly` property performing a virtual function call to a function not present in any of the specified sets.
- Kernel with `calls_indirectly` property performing a virtual function call to a function not marked as indirectly callable.
- Kernel with `calls_indirectly` property performing a virtual function call to a function marked as indirectly callable only in declaration.
- Kernel with `calls_indirectly` property performing a virtual function call to a function marked as indirectly callable only in definition.

Reply:
All of these require the FE to parse property list which we don't really want to do. 

Leaving that aside for now though, I THINK diagnostic 1 can be implemented. Diagnostic 3 is tricky because FE will not know which function is actually called. The visitor will always visit virtual methods based on type of object irrespective of what it is actually pointing to. I'm not sure about 2,4 and 5 off the top of my head.

Again though, FE currently cannot recognize specific properties and we should aim to keep it that way.

Reply:
> Leaving that aside for now though, I THINK diagnostic 1 can be implemented

I think it is important to have this diagnostic.  Note that the compiler already produces a diagnostic if a kernel attempts to make a virtual function call.  Therefore, if we do not have diagnostic 1, that means we will have a regression -- the compiler will no longer diagnose an error condition that it used to diagnose.

Perhaps this is a reason why our design should allow the CFE to recognize a small subset of the properties.

Reply:
> Perhaps this is a reason why our design should allow the CFE to recognize a small subset of the properties.

I would suggest us to turn virtual functions property into a separate attribute then. "allow CFE to recognize a small subset of the properties"  opens up a flood gate and I'm not sure if we would be able to keep that "small subset" really small. Plus, it kind of defeats the whole purpose of generic (to the CFE, at least) library-friendly properties.

Moreover, parsing properties arguments will be less efficient, because they are stored as plain strings and we would have to scan all of them over and over in every place we want to check semantics based on properties. Attributes are encoded into AST and we can use less expensive checks for their presence. Optimizing storage for compile-time properties would likely mean just implementing attributes infrastructure within a single attribute, which doesn't seem reasonable.

Reply:
Diagnostic (1) can actually be implemented in a middle-end pass: we can parse properties in there since they will already be converted into function attributes/metadata and we can detect indirect calls.

The only complication of doing so in ME is that it will likely be impossible to distinguish between a call to virtual function and a call through a regular function pointer stored somewhere (in general case, I mean). A bit hacky, but we can emit some extra markup using instructions metadata to help ME pass differentiate virtual function calls from other types of indirect calls.